test/integration/quic_http_integration_test.cc

```diff
@@ -1167,6 +1167,86 @@ TEST_P(QuicHttpIntegrationTest, MultipleNetworkFilters) {
+  std::string protocol = log.substr(0, log.find(delimiter));
+  std::string roundtrip_duration = log.substr(log.find(delimiter) + 1, log.length());
+  EXPECT_EQ(protocol, "HTTP/3");
+  EXPECT_GT(std::stoi(roundtrip_duration), 0);
+}
+
+TEST_P(QuicHttpIntegrationTest, DeferredLoggingDisabled) {
+  config_helper_.addRuntimeOverride("envoy.reloadable_features.quic_defer_logging_to_ack_listener",
+                                    "false");
+  useAccessLog("%PROTOCOL%,%ROUNDTRIP_DURATION%");
+  initialize();
+  codec_client_ = makeHttpConnection(makeClientConnection(lookupPort("http")));
+  sendRequestAndWaitForResponse(default_request_headers_, /*request_size=*/0,
+                                default_response_headers_,
+                                /*response_size=*/0,
+                                /*upstream_index=*/0, TestUtility::DefaultTimeout);
+  codec_client_->close();
+
+  // Do not flush client acks.
+  std::string log = waitForAccessLog(access_log_name_, 0, false, nullptr);
+  std::string delimiter = ",";
+  std::string protocol = log.substr(0, log.find(delimiter));
+  std::string roundtrip_duration = log.substr(log.find(delimiter) + 1, log.length());
+  EXPECT_EQ(protocol, "HTTP/3");
+  EXPECT_EQ(roundtrip_duration, "-");
+}
+
+TEST_P(QuicHttpIntegrationTest, DeferredLoggingWithReset) {
+  useAccessLog("%PROTOCOL%,%ROUNDTRIP_DURATION%");
+  initialize();
+  codec_client_ = makeHttpConnection(makeClientConnection(lookupPort("http")));
+  auto response = codec_client_->makeHeaderOnlyRequest(default_request_headers_);
+  waitForNextUpstreamRequest(0);
+  codec_client_->close();
+  ASSERT_TRUE(response->waitForReset());
+  EXPECT_FALSE(response->complete());
+
+  std::string log = waitForAccessLog(access_log_name_);
+  std::string delimiter = ",";
+  std::string protocol = log.substr(0, log.find(delimiter));
+  std::string roundtrip_duration = log.substr(log.find(delimiter) + 1, log.length());
+  EXPECT_EQ(protocol, "HTTP/3");
+  EXPECT_EQ(roundtrip_duration, "-");
+}
+
+TEST_P(QuicHttpIntegrationTest, DeferredLoggingWithQuicReset) {
+  useAccessLog("%PROTOCOL%,%ROUNDTRIP_DURATION%");
+  initialize();
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+
+  // omit required authority header to invoke EnvoyQuicServerStream::resetStream
+  auto encoder_decoder = codec_client_->startRequest(Http::TestRequestHeaderMapImpl{
+      {":method", "GET"}, {":path", "/dynamo/url"}, {":scheme", "http"}});
+  request_encoder_ = &encoder_decoder.first;
+  auto response = std::move(encoder_decoder.second);
+
+  ASSERT_TRUE(response->waitForEndStream());
+  codec_client_->close();
+  ASSERT_TRUE(response->complete());
+
+  std::string log = waitForAccessLog(access_log_name_);
+  std::string delimiter = ",";
+  std::string protocol = log.substr(0, log.find(delimiter));
+  std::string roundtrip_duration = log.substr(log.find(delimiter) + 1, log.length());
+  EXPECT_EQ(protocol, "HTTP/3");
+  EXPECT_EQ(roundtrip_duration, "-");
+}
+
```

Comment:
Please also add a test case where the client reset the request after fake_upstream_ has finish sending response to verify that log_handlers log() is called all the time.

Reply:
I thought that the existing `DeferredLoggingWithReset` covered this case:

```
  waitForNextUpstreamRequest(0);
  codec_client_->close();
```

Can you clarify if I'm missing something?

Reply:
`waitForNextUpstreamRequest(0)` doesn't guarantee that response has been sent back. The test case that I asked for would be able to verify that the deferred logging should be triggered by stream reset while waiting for the ACK.

Reply:
I've added a fix and test for this in https://github.com/envoyproxy/envoy/commit/9b31dcd30db1e24729eed8d7d1777cb030f21533.